src/function/gds/strongly_connected_components.cpp

```diff
@@ -0,0 +1,250 @@
+#include "binder/expression/expression_util.h"
+#include "common/types/types.h"
+#include "function/gds/auxiliary_state/gds_auxilary_state.h"
+#include "function/gds/gds_frontier.h"
+#include "function/gds/gds_function_collection.h"
+#include "function/gds/gds_object_manager.h"
+#include "function/gds/gds_utils.h"
+#include "function/gds/output_writer.h"
+#include "function/gds_function.h"
+#include "graph/graph.h"
+#include "processor/execution_context.h"
+#include "processor/result/factorized_table.h"
+
+using namespace std;
+using namespace kuzu::binder;
+using namespace kuzu::common;
+using namespace kuzu::processor;
+using namespace kuzu::storage;
+using namespace kuzu::graph;
+
+namespace kuzu {
+namespace function {
+
+offset_t NOT_ASSIGNED = std::numeric_limits<offset_t>::max();
+
+/// State for the SCC algorithm
+class SCCState {
+public:
+    SCCState(const table_id_map_t<offset_t>& numNodesMap, MemoryManager* mm) {
+        for (const auto& [tableID, numNodes] : numNodesMap) {
+            componentIDsMap.allocate(tableID, numNodes, mm);
+            visitsMap.allocate(tableID, numNodes, mm);
+            pinTable(tableID);
+            for (auto i = 0u; i < numNodes; ++i) {
+                visits[i].store(false, std::memory_order_relaxed);
+                componentIDs[i].store(NOT_ASSIGNED, std::memory_order_relaxed);
+            }
+        }
+    }
+
+    void pinTable(table_id_t tableID) { componentIDs = componentIDsMap.getData(tableID); 
+        visits = visitsMap.getData(tableID);}
+
+    bool updateVisit(common::offset_t offset, bool value) {
+        visits[offset].store(value, std::memory_order_relaxed);
+        return true;
+    }
+
+    bool getVisit(offset_t offset) {
+        return visits[offset].load(std::memory_order_relaxed);
+    }
+
+    bool updateComponentID(common::offset_t offset, offset_t value) {
+        componentIDs[offset].store(value, std::memory_order_relaxed);
+        return true;
+    }
+
+    offset_t getComponentID(offset_t offset) {
+        return componentIDs[offset].load(std::memory_order_relaxed);
+    }
+
+private:
+    // Pointers to current tableID's visits and componentIDs
+    std::atomic<bool>* visits = nullptr;
+    std::atomic<offset_t>* componentIDs = nullptr;
+    // State for visits and componentIDs
+    ObjectArraysMap<std::atomic<bool>> visitsMap;
+    ObjectArraysMap<std::atomic<offset_t>> componentIDsMap;
+};
+
+class SCCCompute {
+public:
+    SCCCompute(graph::Graph* graph, const table_id_map_t<offset_t>& numNodesMap,
+        SCCState& sccState)
+        : graph{graph}, sccState{sccState}, numNodesMap{numNodesMap} {};
+    ~SCCCompute() = default;
+
+    // Forward DFS
+    void visit_vertex(nodeID_t node, catalog::TableCatalogEntry* relEntry) {
+        if (sccState.getVisit(node.offset)) {
+            // Already visited
+            return;
+        };
+        sccState.updateVisit(node.offset, true);
+        auto scanState = graph->prepareRelScan(relEntry, "");
```

Comment:
This state should be recreated in each recursion, right?

Reply:
So I implemented this if I understood you correctly. Is it the case that maintaining a second copy of the graph is faster than scanning the main graph?